### ISSUE TITLE
using connection.query(...).stream() with promise wrapper

### ISSUE THREAD

**Issue by jomel** [sidorares/node-mysql2#677]
why can't I use `connection.query(...).stream() ` with promise wrapper?
(you get error: TypeError: connection.query(...).stream is not a function)

I'd quite like to be able to do: 
`let q = await connection.query(someSQL);`
as well as      
```
const s1 = connection.query(veryBiqSqlResult).stream();
s1.on('result', async function(row) {...})
```
in the same script, but atm I need promise api for the first and standard api nor the second...

**Comment by sidorares** [comment]
maybe we need to wait for async generators to land: http://node.green/#ESNEXT-candidate--stage-3--Asynchronous-Iterators-for-await-of-loops

```js
for await (const row of connection.query(veryBiqSqlResult).stream()) {
  console.log(row);
}
```
( not sure if I can mix buffering `.query() => Promise version` with non-buffering detecting that .stream() is added, maybe we need another name: `for await (const row of connection.queryStream(veryBiqSqlResult)) {})

You can access non-promise api from wrapper as `connection.connection` - see https://github.com/sidorares/node-mysql2/blob/97a88530db72c656c16f071a710701ad2830b38d/promise.js#L55

```js
const [rows] = await connection.query(smallSqlResult);
await new Promise((accept, reject) => {
  const s1 = connection.connection.query(veryBiqSqlResult).stream();
  s1.on('result', function(row) {...})
  s1.on('end', accept);
  s1.on('error', reject);
})
```

**Comment by derN3rd** [comment]
One note to the connection.connection.query(...).stream() part:
I had problems with it, until I saw the comment https://github.com/sidorares/node-mysql2/issues/770#issuecomment-381295768
After removing .stream() it worked perfectly

**Comment by sidorares** [comment]
Oh, thanks @derN3rd I'll update my example here

**Comment by sidorares** [comment]
yes, `stream()` is just to wrap into pipeable object, and _not_ to enable 'result' events, they are emitted anyway

**Comment by terrisgit** [comment]
```
s1.on('result', async function(row) {...})
```

I tried passing to 'on()' a function that returns a promise. The result appears to be that the rows are processed in random order. In other words, it appears that connection.query is not awaiting on these handlers, which is something I need. Or maybe my testing isn't working correctly.

**Comment by terrisgit** [comment]
Don't forget to call connection.connection.release when you're done or you'll get crashes if the SQL statement contains syntax errors and, I think, also mysql connection leaks.

**Comment by terrisgit** [comment]
This is probably a problem in node-mysql, but if you kill the database while processing a large rowset, neither end() nor error() are called. I'll try to create a test case.

**Comment by sidorares** [comment]
@terrisgit try setting `enableKeepAlive` option to `true` - see discussion in https://github.com/sidorares/node-mysql2/pull/1081

Maybe we should change that to be true by default, but not sure if there might be some undesired side effects

**Comment by terrisgit** [comment]
end() and error() still aren't getting called. I'll create an issue.

**Comment by terrisgit** [comment]
Replied too soon. It seemed to be working, but it stopped working. No idea what I did. end() and error() are still not getting called. I'll create a small example.

**Comment by terrisgit** [comment]
This is the only semi-complete example on the WWW that I know of that creates a CSV file efficiently in JavaScript and correctly handles loss of connection and syntax errors. Flushing the outputStream at the end is a little tricky but that's not the hard part.

```
const csvstringify = require('csv-stringify');
const fs = require('fs');

const BOM = '\ufeff';
const outputStream = fs.createWriteStream('output.csv', {encoding: 'utf8'});

// const connection = __Create a mysql2 Connection object here__
const generator = connection.connection.query('select 1');
let recordsProcessed = 0;

generator
  .on('fields', fields=>{
    outputStream.write(BOM+fields.map(field=>field.name).join(',')+'\n');
  })
  .on('result', row=>{
    ++recordsProcessed; // Counting rows just as an example
  });

try {
  await new Promise((resolve, reject) => {
    // This works only when enableKeepAlive is true for the connection pool
    // See https://github.com/sidorares/node-mysql2/issues/677#issuecomment-588530194
    connection.on('error', error=>{
      reject(error);
    });

    generator
      .stream({highWaterMark: this.highWaterMark})
      .on('error', reject)
      .on('end', resolve)
      .pipe(csvstringify())
      .pipe(outputStream);
  });
}
finally {
  connection.on('error', function() {});
}
```

**Comment by terrisgit** [comment]
I've been doing a lot of error testing. The above code can be forced to fail in many ways but I've been focusing on write errors which you can force via outputStream.close() after you create it. It seems to leak memory if you run the above code in a loop, even if you force garbage collection. It's not a typical runtime error anyway so it's not a concern.

**Comment by eyalroth** [comment]
I ended up with this workaround, if someone's interested:

```js
const unpromisedConnection = connection.connection;
const queryUnpromised = unpromisedConnection.query.bind(unpromisedConnection);

connection.streamQuery = async (sql, values, streamOptions) =>
  new Promise((resolve, reject) => {
    const query = queryUnpromised(sql, values);
    const stream = query.stream(streamOptions);

    stream.once('error', reject);
    query.once('fields', () => {
      stream.removeListener('error', reject);
      resolve(stream);
    });
    query.once('error', reject);
  });
```

This adds the method `streamQuery` to the connection which accepts the same arguments as `connection.query(sql, values).stream(streamOptions)`, and returns a promise that is resolved with the stream once the result set returns from the connection *before* the first result arrives, or rejected if the query fails.

The resolved stream may still emit `error` events IIUC, so those should still be handled by the caller:

```js
connection.streamQuery('SELECT blabla malformed sql'); // promise reject

const stream = await connection.streamQuery('SELECT name FROM employees');
stream.on('error', console.log);
stream.pipe(...);
```

**Comment by eyalroth** [comment]
Following problems with error handling that I've encountered while trying to use queue composition and promises/async functions -- described in https://github.com/nodejs/node/issues/39722 -- I feel like it would be unsafe to add any kind of support for promisified queue API, other than perhaps injecting the `queue` method to the promise returned from the `query` method if that's possible.

**Comment by marnixhoh** [comment]
Just a note for anyone running into this issue. I had to actually include .stream() to make it work, as opposed to not calling it as suggested in this comment:

> One note to the connection.connection.query(...).stream() part: I had problems with it, until I saw the comment [#770 (comment)](https://github.com/sidorares/node-mysql2/issues/770#issuecomment-381295768) After removing .stream() it worked perfectly

My code ended up looking something like this:

```
connection.connection.query('SELECT * FROM my_table')
    .stream()
    .pipe(someStream)
```

2 things to note:
1. `connection.connection` is used as opposed to just `connection` as this accesses the non-promise API.
2. It is not possible to stream individual columns, therefor. you must use `*` to get all of them. For more info: [see the docs](https://github.com/mysqljs/mysql#streaming-query-rows)

**Comment by eyalroth** [comment]
> 2. It is not possible to stream individual columns, therefore you must use `*` to get all of them. For more info: [see the docs](https://github.com/mysqljs/mysql#streaming-query-rows)

Interesting. I wonder whether this works:

```js
connection.connection.query('SELECT * FROM (SELECT column_a FROM my_table)')
    .stream()
    .pipe(someStream)
```

**Comment by marnixhoh** [comment]
@eyalroth I have no idea... I think the docs are a little unclear on this. In case you're interested, I have asked a somewhat similar question here: https://github.com/mysqljs/mysql/issues/2523

**Comment by marnixhoh** [comment]
@eyalroth I just wanted to let you (and others) know that I misinterpreted the docs. I have edited my comment accordingly.

It is indeed possible to stream individual columns. However, it is not possible to stream individual fields. So a large field/entry can not be streamed and will be buffered in its entirety. I hope this clarifies it. [More info](https://github.com/mysqljs/mysql/issues/2523)
